File: FSharpLib/Geometry/Box.fs
namespace Lib.Geometry

module Box =

    // Represents a three-dimensional box in a world where boxes may overlap
    type Box = Box of int * int * int * int * int * int with
    
        // The box representing the overlap of b1 and b2
        static member overlap b1 b2 =
            let (minX1, maxX1, minY1, maxY1, minZ1, maxZ1) = b1
            let (minX2, maxX2, minY2, maxY2, minZ2, maxZ2) = b2
            let minX = max minX1 minX2
            let maxX = min maxX1 maxX2
            let minY = max minY1 minY2
            let maxY = min maxY1 maxY2
            let minZ = max minZ1 minZ2
            let maxZ = min maxZ1 maxZ2
            if minX > maxX || minY > maxY || minZ > maxZ
            then None
            else Some (minX, maxX, minY, maxY, minZ, maxZ)
    
        // All boxes representing overlap with the first box
        static member totalOverlaps box boxes = 
            let rec tOverlaps box boxes = 
                match boxes with 
                | b :: bs -> match Box.overlap box b with
                             | Some bb -> bb :: tOverlaps box bs
                             | _       -> tOverlaps box bs
                | _       -> []
            tOverlaps box boxes

        // The volume of a box
        static member volume b =
            let (minX, maxX, minY, maxY, minZ, maxZ) = b
            bigint (maxX + 1 - minX) * bigint (maxY + 1 - minY) * bigint (maxZ + 1 - minZ)

        // The total volume of all boxes, taking into account that they may overlap
        static member totalVolume boxes =
            let rec tVolume boxes =
                match boxes with
                | b :: bs -> let overlaps = Box.totalOverlaps b bs
                             Box.volume b + tVolume bs - tVolume overlaps
                | _       -> bigint 0
            tVolume boxes

File: CSharpLib/DataStructures/Box.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CSharpLib.DataStructures
{
    public class Box
    {
        public long MinX { get; }
        public long MinY { get; }
        public long MinZ { get; }
        public long MaxX { get; }
        public long MaxY { get; }
        public long MaxZ { get; }

        public Box(long minX, long minY, long minZ, long maxX, long maxY, long maxZ)
        {
            MinX = minX;
            MinY = minY;
            MinZ = minZ;
            MaxX = maxX;
            MaxY = maxY;
            MaxZ = maxZ;
        }

        /// <summary>
        /// Return the volume of this box
        /// </summary>
        public long Volume()
        {
            return (MaxX - MinX + 1) * (MaxY - MinY + 1) * (MaxZ - MinZ + 1);
        }

        /// <summary>
        /// Return a box representing the volume which both this and 
        /// the other box occupies. If no volume is occupied by both 
        /// boxes, returns null.
        /// </summary>
        public Box? Intersection(Box other)
        {
            var minX = Math.Max(MinX, other.MinX);
            var minY = Math.Max(MinY, other.MinY);
            var minZ = Math.Max(MinZ, other.MinZ);
            var maxX = Math.Min(MaxX, other.MaxX);
            var maxY = Math.Min(MaxY, other.MaxY);
            var maxZ = Math.Min(MaxZ, other.MaxZ);
            if (minX > maxX || minY > maxY || minZ > maxZ)
            {
                return null;
            }
            return new Box(minX, minY, minZ, maxX, maxY, maxZ);
        }

        /// <summary>
        /// Return all non-null intersections with this box.
        /// </summary>
        public IEnumerable<Box> Intersections(IEnumerable<Box> others)
        {
            foreach (var box in others)
            {
                var overlap = Intersection(box);
                if (overlap != null)
                {
                    yield return overlap;
                }
            }
        }

        /// <summary>
        /// Return the total volume of all boxes, taking 
        /// into account that they may overlap
        /// </summary>
        public static long TotalVolume(IEnumerable<Box> boxes)
        {
            if (boxes.Count() == 0)
            {
                return 0;
            }

            var thisBox = boxes.First();
            var otherBoxes = boxes.Skip(1);

            if (boxes.Count() == 1)
            {
                return thisBox.Volume();
            }

            return thisBox.Volume() + TotalVolume(otherBoxes) - TotalVolume(thisBox.Intersections(otherBoxes));
        }
    }
}
